Add Flint matrices for ZZ and QQ

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Followup from gh-25474

Built on top of gh-25492 (which could be merged first or otherwise this could be merged instead).

See also gh-25410.

#### Brief description of what is fixed or changed

Uses python-flint's fmpz and fmpq if `SYMPY_GROUND_TYPES=flint` for dense matrices over ZZ and QQ.

#### Other comments

Plenty more work can be done to make this faster but first step is to get it working.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
    * The dense implementation of DomainMatrix will now use python-flint matrices for matrices over ZZ and QQ if `SYMPY_GROUND_TYPES=flint` and python-flint is installed. A new low-level matrix type `DFM` is added to be a wrapper for dense flint matrices. `DomainMatrix` will use this internally by default for dense matrices when the ground types are `flint`. This makes many operations with dense DomainMatrix much faster although for now to test this you need to explicitly choose to use the dense implementation by doing e.g. `M = Matrix([[1, 2], [3, 4]]).to_DM().to_dense()` to get a matrix that has a flint matrix as its internal representation.
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
test(polys): Add CI tests for GROUND_TYPES=flint

oscarbenjamin committed:
test(polys): fix doctests for flint compatibility

oscarbenjamin committed:
Update taball list and flake8

oscarbenjamin committed:
Fix typo

oscarbenjamin committed:
feat(matrices): add FDM for Flint matrices

oscarbenjamin committed:
Add matrices to the flint tests

oscarbenjamin committed:
test(CI): skip DFM doctests if python-flint not installed

oscarbenjamin committed:
doc(polys): Fix the docs build for DFM

smichr commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -158,17 +162,22 @@ def __new__(cls, rows, shape, domain, *, fmt=None):
+
         return cls.from_rep(rep)
 
-    def __getnewargs__(self):
```
This was deleted/renamed -- what is using this? I don't see this anywhere else in the diff and don't see use of `reduce(`, either.

oscarbenjamin replied:
This was changed to `__reduce__`. The problem with `__getnewargs__` is that pickle still looks in `__dict__` and finds the `.rep` attribute which is a flint matrix. Flint matrices are not picklable so that fails.

Instead we need to implement `__reduce__` which gives full control over what objects are used for pickling. In this case we can just convert to list of lists format because flint's `fmpz` and `fmpq` are picklable.

oscarbenjamin committed:
test(matrices): add unified tests for DDM/DFM/SDM

smichr commented on sympy/polys/matrices/ddm.py:
```diff
@@ -101,13 +101,18 @@ class DDM(list):
     is_DDM = True
 
     def __init__(self, rowslist, shape, domain):
+        if not (isinstance(rowslist, list) and all(type(row) is list for row in rowslist)):
```
If the number of columns is 1 can it just be a list as for matrix (instead of a list of singleton lists)?

oscarbenjamin replied:
I prefer not to add those kinds of complications at this level. The constructor here is not really expected to be something that would be used directly by users and so it is better to be strict about what is expected for inputs: prioritise speed and simplicity of implementation over convenience of use.

The convenience of use way leads to `_handle_creation_inputs`:
https://github.com/sympy/sympy/blob/b7a7ee1b7096486d8ecd460332506e04033cce13/sympy/matrices/matrices.py#L896-L1118
I hope I never have to touch that function again!

oscarbenjamin replied:
Oh, btw if you are looking to create a DDM conveniently then it is usually better to convert from something else e.g.:
```python
In [3]: M = Matrix([[1, 2], [3, 4]])

In [4]: M.to_DM() # DomainMatrix
Out[4]: DomainMatrix({0: {0: 1, 1: 2}, 1: {0: 3, 1: 4}}, (2, 2), ZZ)

In [5]: M.to_DM().to_ddm() # DDM
Out[5]: DDM([[mpz(1), mpz(2)], [mpz(3), mpz(4)]], (2, 2), ZZ)

In [6]: M.to_DM().to_sdm() # SDM
Out[6]: SDM({0: {0: mpz(1), 1: mpz(2)}, 1: {0: mpz(3), 1: mpz(4)}}, (2, 2), ZZ)

In [7]: M.to_DM().to_dfm() # DFM (needs ground types = flint)
...
NotImplementedError: DFM requires GROUND_TYPES=flint.
```

oscarbenjamin committed:
feat(polys): avoid conversions in some DFM methods

oscarbenjamin committed:
perf(polys): Use the python-flint det if possible

oscarbenjamin commented:
In the last commit I have changed it so that `DFM.det` calls through to the FLINT implementation of either `fmpz_mat_det` or `fmpq_mat_det`.

There is no sparse implementation for `SDM.det` which just converts to dense and calls `DDM.det`. I have changed that here to convert to `DFM` if possible so that now sparse domain matrices will end up using the Flint `det` method for matrices over ZZ or QQ. That makes computing determinants much faster:
```python
In [3]: M = randMatrix(1000) # 1000x1000 matrix

In [4]: %time ok = M.to_DM().det() # Using ground types = flint
CPU times: user 18.9 s, sys: 74.5 ms, total: 19 s
Wall time: 19 s
```

oscarbenjamin committed:
perf(matrices): Use python-flint inv if possible

oscarbenjamin commented:
The last commit uses python-flint for the inverse of matrices over QQ in DFM. As for `det` the SDM method just calls the DDM method and so is changed here to call the DFM method if possible.

This makes computing the inverse of DomainMatrix over QQ much faster:
```python
In [19]: M = randMatrix(200) # 200x200 matrix

In [20]: %time ok = M.to_DM().to_dense().to_field().inv()
CPU times: user 3.41 s, sys: 15.9 ms, total: 3.42 s
Wall time: 3.41 s
```
Without flint (but with gmpy2) the existing DomainMatrix implementation gives
```python
In [7]: M = randMatrix(200) # 200x200 matrix

In [8]: %time ok = M.to_DM().to_dense().to_field().inv()
CPU times: user 1min 57s, sys: 51.1 ms, total: 1min 57s
Wall time: 1min 57s
```
The difference is firstly algorithmic because fraction-free methods are better in this case so DomainMatrix `inv_den` is a lot faster:
```python
In [10]: %time ok = M.to_DM().inv_den()
CPU times: user 14.2 s, sys: 0 ns, total: 14.2 s
Wall time: 14.2 s
```
Compared to Flint that is about 4x slower which is consistent with other comparisons of Flint vs DomainMatrix fraction-free methods. I believe that 4x difference is primarily due to some algorithmic difference in solving linear systems like maybe Flint uses LU solve whereas DomainMatrix always uses RREF for solving.

oscarbenjamin committed:
perf(matrices): Use python-flint inv if possible

oscarbenjamin commented:
The last commit enables using flint for charpoly if possible. Same story as before: there is no sparse implementation and so `SDM` uses `DDM.charpoly`. Here that is changed to use `DFM.charpoly` if possible which in turn calls into the Flint routine.

Note that I have a PR gh-25458 which adds a sparse implementation of `charpoly`. The two PRs are compatible but the method for selecting an algorithm in gh-25458 should probably always choose Flint when possible. The Flint algorithm is modular with CRT reconstruction and is significantly better than any algorithms currently in SymPy.

Timing:
```python
In [1]: M = randMatrix(200) # 200x200 matrix

In [2]: %time ok = M.to_DM().charpoly()
CPU times: user 1.49 s, sys: 81 µs, total: 1.49 s
Wall time: 1.49 s
```
Without Flint that example is about 60x slower:
```python
In [11]: M = randMatrix(200) # 200x200 matrix

In [12]: %time ok = M.to_DM().charpoly()
CPU times: user 1min 26s, sys: 43.5 ms, total: 1min 26s
Wall time: 1min 26s
```
The difference is algorithmic though so I expect the ratio to grow for larger matrices.

oscarbenjamin committed:
perf(matrices): use flint method for to_list_flat

oscarbenjamin committed:
perf(matrices): Use Flint for charpoly and lu_solve

oscarbenjamin committed:
perf(matrices): use flint for LLL

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_flint_matrix

oscarbenjamin committed:
fix(polys): remove unneeded imports

oscarbenjamin committed:
matrices: add TODO comment

oscarbenjamin committed:
skip doctests when flint not installed

oscarbenjamin commented:
This is ready for review now.

I was originally thinking that the much simpler PR gh-25492 might be merged first just because it is simpler and sort of independent of the other changes here. It doesn't matter too much but I have already needed to fix merge conflicts in gh-25492 so it would be good to get that in rather than leaving it open.

This PR adds a `DFM` class to parallel the existing `DDM` and `SDM` classes that are the dense and sparse internal implementations of `DomainMatrix`. The `DFM` class wraps a Flint matrix which for now is either an `fmpz_mat` (a dense integer matrix) or an `fmpq_mat` (a dense rational matrix). Since Flint is implemented in C and has good algorithms for all of the operations used here this can bring some significant speedups.

This is still just a start in terms of making use of python-flint in SymPy. I could open many more PRs like this but I have a lot of open PRs right now so I'm going to wait for some to get reviewed or merged before opening any more.

To test the PR you should install `python-flint`:
```console
pip install python-flint
```
and then set the `SYMPY_GROUND_TYPES=flint` environment variable as described above. Here are some timings.

Matrix multiplication:
```python
In [25]: M = randMatrix(100) # 100x100 matrix of smallish integers

In [26]: dM = M.to_DM().to_dense() # Convert to a dense DomainMatrix

In [27]: type(dM.rep) # The DFM type added in this PR
Out[27]: sympy.polys.matrices._dfm.DFM

In [28]: type(dM.rep.rep) # The underlying Flint type
Out[28]: flint._flint.fmpz_mat

In [29]: %time ok = dM * dM # time to multiply two 100x100 Flint matrices
CPU times: user 4.64 ms, sys: 4 µs, total: 4.64 ms
Wall time: 4.79 ms

In [30]: %time ok = M * M # time to multiply with SymPy pure Python implementation
CPU times: user 458 ms, sys: 4.01 ms, total: 462 ms
Wall time: 461 ms
```
So the difference here is that Flint can multiply two dense integer matrices about 100x faster than SymPy's existing implementation. Note that for this operation `DomainMatrix` is used internally so this really comparing `DomainMatrix` against Flint. The factor of 100 difference is typical of optimised Python code vs optimised C code. For larger matrices the difference is larger and then the difference is algorithmic:
```python
In [31]: M = randMatrix(200) # 200x200 matrix of smallish integers

In [32]: dM = M.to_DM().to_dense() # Convert to a dense DomainMatrix

In [33]: %time ok = dM * dM # time to multiply two 200x200 Flint matrices
CPU times: user 11.6 ms, sys: 0 ns, total: 11.6 ms
Wall time: 11.5 ms

In [34]: %time ok = M * M # time to multiply with SymPy pure Python implementation
CPU times: user 3.38 s, sys: 8 µs, total: 3.38 s
Wall time: 3.38 s
```
Here for 200x200 matrices we see that Flint is about 300x faster. The difference in ratio here is because the underlying algorithm is different (I'm not sure exactly what the algorithmic difference is yet). We can expect that for large matrices that difference in timing will grow.

This PR means that in principle SymPy could use Flint for operations like multiplication although the PR does not actually enable that. Deciding whether or not to use Flint for a given operation is complicated because the existing SymPy implementation for things like this uses a sparse DomainMatrix representation which is optimised for sparse matrices. If we make the matrices more and more sparse then at some point the existing sparse DomainMatrix implementation will be faster than Flint's dense representation:
```python
In [35]: M = randMatrix(200, percent=1) # 200x200 matrix with 1% entries nonzero

In [36]: dM = M.to_DM().to_dense() # Convert to a dense DomainMatrix

In [37]: %time ok = dM * dM # time to multiply two 200x200 sparse Flint matrices
CPU times: user 24.3 ms, sys: 2 µs, total: 24.3 ms
Wall time: 24.1 ms

In [38]: %time ok = M * M # time to multiply with SymPy pure Python implementation
CPU times: user 2.67 ms, sys: 0 ns, total: 2.67 ms
Wall time: 2.69 ms
```
The situation in master right now is that SymPy's current sparse representation is always faster than its dense representation so we do not need to have any code that would decide which representation to use: just always use the sparse representation. After this PR the situation would be more complicated and we would need to have code that checks exactly how sparse the matrix is to decide whether to use the sparse implementation or to use Flint if it is available. I have not added that here so as of this PR the sparse representation is still used by default except in cases that already convert to the dense representation (e.g. because a sparse implementation of an algorithm is not implemented such as the operations mentioned below).

In future the approach here could be used to add support for many more types of matrices but for now this only applies to matrices with integer or rational entries. If `SYMPY_GROUND_TYPES=flint` then this PR immediately speeds up a number of operations for `DomainMatrix`:

- det
- charpoly
- inv
- lu_solve
- lll

The differences can be very significant because Flint is faster both as a result of being implemented in C rather than Python and also because it generally uses the best algorithms (at least better than many of SymPy's existing algorithms). In some cases the SymPy algorithms could be improved to give performance that more closely resembles the Flint performance. In other cases like the matrix multiplication example above the Python overhead is dominant and a pure Python implementation could never become close in speed to a C implementation. In any case Flint is the state of the art for these kinds of exact/symbolic calculations so being able to use them makes SymPy as fast as anything else. Note that other computer algebra systems (Mathematica etc) already use Flint for exactly this reason.

There are more operations that are implemented in python-flint that I have not immediately added in this PR for various reasons. The PR is already large and for some of the operations that it would be good to add the best thing would be to make some nontrivial changes to the existing DomainMatrix classes. I would prefer to do that in separate PRs that can be reviewed independently more easily then the changes here.

Some examples of things that I have not implemented and reasons are:

- Reduced row echelon form (RREF). The flint matrices do support  RREF and it is fast. However we now have a more complicated method for selecting RREF algorithms and so if we add support for Flint's RREF algorithms then it needs to be incorporated into that. Having worked on this PR I am now less happy with the compromises I agreed to in gh-25443 but I would prefer to discuss that in a separate PR from this one.
- Smith normal form (SNF) and Hermite normal form (HNF). The flint integer matrices have this and it is fast. However the existing DDM and SDM types do not have an `snf` method and the existing code works only with DomainMatrix as input rather than DDM. This is not difficult to change but it requires a small refactor that I want to avoid in this PR.
- Nullspace. There are some problems with the signature of the DomainMatrix, DDM and SDM nullspace methods that make it difficult to use Flint here. That is not hard to fix but again requires a small refactor that I want to avoid in this PR. Also in python-flint only matrices over ZZ have a nullspace method and not matrices over QQ which is the reverse of the way that DomainMatrix currently works. The nullspace method should be added for QQ in python-flint. Also the DDM and SDM nullspace methods should be changed to work over rings rather than just fields.
- Certain DomainMatrix (and DDM and SDM) methods require a field and will raise an error otherwise e.g. `inv`, `solve`, `rref`, `lu`, `lu_solve`. The corresponding python-flint matrix methods will just convert to a field so e.g. `fmpz_mat.inv` will just return a `fmpq_mat`. The `fmpz_mat.inv` method has an `integer=True` option that will compute the inverse of an integer matrix as an integer matrix if possible (if the matrix is unimodular) and raise otherwise. I think that I prefer the python-flint approach here and DomainMatrix should generally move to that instead. Again this is a change that I don't want to make in this PR so that it can be discussed separately.

I have also added a test module that uses pytest's parametrised test functionality to run the same test suite across all of SDM, DDM and the new DFM introduced here. Running the same tests on all implementations reveals some inconsistencies. Since these types are supposed to be interchangeable I have made some small changes to the existing types to improve compatibility but I have deliberately left out cases where larger changes would be needed or new methods would need to be added to the existing types (e.g. `hnf` or `snf`).

In future work a lot can be done to expose more of Flint's functionality in python-flint and expose more of python-flint's functionality in DomainMatrix. The final step in each case is exposing DomainMatrix's functionality and performance in Matrix which I am *not* doing in this PR. I *think* that none of the changes here will impact anything with `Matrix` because the relevant methods are not used by `Matrix` (also the new class introduced here is only use if `SYMPY_GROUND_TYPES=flint`).

I have separate PRs to make Matrix use DomainMatrix e.g. gh-25458 (charpoly), gh-25452 (inv) but the only one that is merged is gh-25443 (RREF) and RREF is one of the operations that I am *not* changing here. I could easily open more PRs like that but I am struggling to get any significant review on the existing open PRs so I won't push further on that right now.

This PR should not adversely affect anything that is directly used by users. It does however add the capability to SymPy to leverage state of the art implementations for many key matrix operations. More work would be needed to translate that into tangible benefits for end users (e.g. like gh-25458 and 25452) but this is an important step. Also there is plenty more in python-flint and in Flint itself that could be leveraged so this should be seen as just a first step in making use of the features there. The biggest benefit to SymPy from Flint would come from Flint's sparse polynomials which are not yet exposed in python-flint but someone else is working on that as I write this and I expect that the next release of python-flint will have something that SymPy can use.

oscarbenjamin commented:
Thanks @smichr!

There was some discussion elsewhere about the requirements file but I'll follow that up in gh-25468
